File: phase-5-roblox/5.3-oop-in-luau/starter/roblox-kingdom/Engine/Building.lua
local Building = {}
Building.__index = Building

function Building.new(name: string)
    local self = setmetatable({}, Building)
    self.name = name
    self.level = 1
    return self
end

function Building:upgrade()
    self.level = self.level + 1
end

function Building:tick(_ledger: any)
    -- default: no production
end

return Building

File: phase-2-persistence/2.3-round-trip-tests/starter/Kingdom.Engine/Buildings/Building.cs
namespace Kingdom.Engine.Buildings;

public class Building
{
    public string Name { get; }
    public int Level { get; private set; } = 1;

    public Building(string name) { Name = name; }

    protected Building(string name, int level) { Name = name; Level = level; }

    public void Upgrade() => Level++;

    public virtual void Tick(ResourceLedger ledger) { }
}
